app/models/work_packages/scopes/allowed_to.rb

```diff
@@ -0,0 +1,184 @@
+      end
+
+      def allowed_to_role_permission_join(permissions) # rubocop:disable Metrics/AbcSize
+        return if permissions.all?(&:public?)
+
+        role_permissions_table = RolePermission.arel_table
+        enabled_modules_table = EnabledModule.arel_table
+        roles_table = Role.arel_table
+
+        condition = permissions.inject(Arel::Nodes::False.new) do |or_condition, permission|
+          permission_condition = role_permissions_table[:permission].eq(permission.name)
+
+          if permission.project_module.present?
+            permission_condition = permission_condition.and(enabled_modules_table[:name].eq(permission.project_module))
+          end
+
+          or_condition.or(permission_condition)
+        end
+
+        arel_table
+          .join(role_permissions_table, Arel::Nodes::InnerJoin)
+          .on(roles_table[:id].eq(role_permissions_table[:role_id])
+                              .and(condition))
+          .join_sources
+      end
+
+      def allowed_to_members_condition(user)
+        members_table = Member.arel_table
+
+        members_table[:project_id].eq(arel_table[:project_id])
+                                  .and(members_table[:user_id].eq(user.id))
+                                  .and(members_table[:entity_type].eq(model_name.name))
+      end
+
+      def allowed_to_builtin_roles_in_active_project_join(user)
+        condition = allowed_to_built_roles_in_active_project_condition(user)
+
+        if user.logged?
+          condition = condition.and(allowed_to_no_member_exists_condition(user))
+        end
+
+        roles_table = Role.arel_table
+
+        arel_table.join(roles_table)
+                  .on(condition)
+                  .join_sources
+      end
+
+      def allowed_to_member_in_active_project_join(user)
+        Project.arel_table
+          .join(Project.arel_table)
+                  .on(Project.arel_table[:active].eq(true)
+                   .and(allowed_to_members_condition(user)))
+                  .join_sources
+      end
+
+      def allowed_to_member_in_work_package_join
+        members_table = Member.arel_table
+        arel_table.join(arel_table)
+        .on(members_table[:entity_id].eq(arel_table[:id]))
+        .join_sources
+      end
+
+      def allowed_to_built_roles_in_active_project_condition(user)
+        projects_table = Project.arel_table
+
+        builtin = if user.logged?
+                    Role::BUILTIN_NON_MEMBER
+                  else
+                    Role::BUILTIN_ANONYMOUS
+                  end
+
+        roles_table = Role.arel_table
+
+        roles_table[:builtin].eq(builtin)
+                             .and(projects_table[:active])
+                             .and(projects_table[:public])
+      end
+
+      def allowed_to_no_member_exists_condition(user)
+        Member
+          .select(1)
+          .where(allowed_to_members_condition(user))
+          .arel
+          .exists
+          .not
+      end
+    end
```

Comment:
This copies a lot of the code from the `Projects::AllowedTo` scope. We'll have to see how much of that is still there after the rework on the work package members check but I suspect that there is still a substantial overlap. Given the complexity and similarity of the use case of the two scopes, let's extract what they share into a module.

Reply:
I have been thinking about this some more. Maybe the better approach is to introduce a scope on Member which allows filtering for all members that grant a permission in some context (project, global, entity). This would exclude the case for non members/anonymous of course but those are rather project specific currently so it might be fine. That scope could then be reused for other entities as well. So ideally, an entity specific scope would just have to be combined out of scopes from the project and members.